### PR TITLE
Incorrect fix in COMPACT_OBJECT_INITIALIZATION rule

### PR DESCRIPTION
**Pull request by kgevorkyan**
### What's done:
* Fix bug and process case, when receiver name is also used in right part of expression
* Add tests

### CODE REVIEW

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter6/classes/CompactInitialization.kt [hunk 1]
```diff
@@ -114,6 +120,14 @@ class CompactInitialization(configRules: List<RulesConfig>) : DiktatRule(
                             }
                             it.treeParent.removeChild(it)
                         }
+                    val receiverName = assignment.text.substringBefore('.')
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter6/classes/CompactInitialization.kt:123]: Maybe, via `psi as KtDotQualifiedExpression`?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter6/classes/CompactInitialization.kt [hunk 2]
```diff
@@ -114,6 +120,14 @@ class CompactInitialization(configRules: List<RulesConfig>) : DiktatRule(
...
+                    // looking for usages of receiver in right part
+                    val identifiers = assignment.right!!.node.findAllDescendantsWithSpecificType(IDENTIFIER)
+                    identifiers.forEach {
+                        if (it.text == receiverName && it.treeParent.treeParent.elementType != CALL_EXPRESSION) {
+                            it.treeParent.replaceChild(it, kotlinParser.createNode("this"))
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter6/classes/CompactInitialization.kt:128]: What type will this node have? Is it simply `THIS_KEYWORD`? Is it correct to substitute `IDENTIFIER` with it?

Comment (kgevorkyan) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter6/classes/CompactInitialization.kt:128]: I did the following:

**1.**  Let's consider the expression, which was obtained from code (nothing changed is AST tree by our hands)
```kotlin
id = executionService.saveExecution(this)
```
Part of AST representation:
```bash
----- CALL_EXPRESSION: "saveExecution(this)"
------ REFERENCE_EXPRESSION: "saveExecution"
------- IDENTIFIER: "saveExecution"
------ VALUE_ARGUMENT_LIST: "(this)"
------- LPAR: "("
------- VALUE_ARGUMENT: "this"
-------- THIS_EXPRESSION: "this"
--------- REFERENCE_EXPRESSION: "this"
---------- this: "this"
------- RPAR: ")"
```


----------------
**2.** Now, let's look at the AST structure when we changed the nodes, and replace `IDENTIFIER` by `this` in:
```kotlin
id = executionService.saveExecution(execution)
```

```bash
-- CALL_EXPRESSION: "saveExecution(this)"
--- REFERENCE_EXPRESSION: "saveExecution"
---- IDENTIFIER: "saveExecution"
--- VALUE_ARGUMENT_LIST: "(this)"
---- LPAR: "("
---- VALUE_ARGUMENT: "this"
----- REFERENCE_EXPRESSION: "this"
------ THIS_EXPRESSION: "this"
------- REFERENCE_EXPRESSION: "this"
-------- this: "this"
---- RPAR: ")"
```

So, yes, **you were right.** There is the difference: we have extra `REFERENCE_EXPRESSION: "this"` node.

Now I changed logic, and looking not for `IDENTIFIER `, but for `REFERENCE_EXPRESSION`, and change it by `this`
In current realization, the AST representation as in **1.** and nothing breaks the structure

**Review by petertrr — APPROVED** [review]
LGTM

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter6/classes/CompactInitialization.kt (hunk 2 above)
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter6/classes/CompactInitialization.kt:128]: Good, thanks for detailed response!

**Pull request merged**